Lower allocations in DirectMemoryAccessStreamReader.ReadToEnd

Neme12 opened the pull request:
The [base `ReadToEnd` implementation](https://referencesource.microsoft.com/#mscorlib/system/io/textreader.cs,144) does some allocations that can be completely avoided if we provide a specialized implementation.

I found this in #61314, but this PR does *not* claim to address the root cause of that issue.

Here's a benchmark showing the reduced allocations (and time) in the new implementation on a 20 kB text.
```
|     Method |                  Job |              Runtime |     Mean |     Error |    StdDev |   Gen 0 |  Gen 1 | Allocated |
|----------- |--------------------- |--------------------- |---------:|----------:|----------:|--------:|-------:|----------:|
|  ReadToEnd |             .NET 6.0 |             .NET 6.0 | 4.433 us | 0.0411 us | 0.0384 us | 14.6332 | 0.0076 |     60 KB |
| ReadToEnd2 |             .NET 6.0 |             .NET 6.0 | 1.359 us | 0.0134 us | 0.0125 us |  4.8771 | 0.0019 |     20 KB |
|  ReadToEnd |        .NET Core 3.1 |        .NET Core 3.1 | 4.993 us | 0.0993 us | 0.1740 us | 14.6484 | 1.8311 |     60 KB |
| ReadToEnd2 |        .NET Core 3.1 |        .NET Core 3.1 | 1.383 us | 0.0100 us | 0.0089 us |  4.8771 | 0.0019 |     20 KB |
|  ReadToEnd | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 4.577 us | 0.0720 us | 0.0674 us | 14.6484 |      - |     60 KB |
| ReadToEnd2 | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.333 us | 0.0157 us | 0.0147 us |  4.8771 |      - |     20 KB |
```

<details>
<summary>Source code of the benchmark</summary>

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using Microsoft.CodeAnalysis.Host;

BenchmarkRunner.Run<ReadToEndBenchmark>();

[MemoryDiagnoser]
[SimpleJob(RuntimeMoniker.Net472)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net60)]
public unsafe class ReadToEndBenchmark
{
    private readonly IntPtr _pointer;
    private readonly TemporaryStorageServiceFactory.DirectMemoryAccessStreamReader _reader1;
    private readonly TemporaryStorageServiceFactory.DirectMemoryAccessStreamReader2 _reader2;

    public ReadToEndBenchmark()
    {
        var length = 10 * 1024;
        _pointer = Marshal.AllocHGlobal(length * sizeof(char));

        _reader1 = new((char*)_pointer, length);
        _reader2 = new((char*)_pointer, length);
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        Marshal.FreeHGlobal(_pointer);
    }

    [Benchmark]
    public string ReadToEnd()
    {
        _reader1.Reset();
        return _reader1.ReadToEnd();
    }

    [Benchmark]
    public string ReadToEnd2()
    {
        _reader2.Reset();
        return _reader2.ReadToEnd();
    }
}
```

Note: for the purpose of this benchmark, I copied the `DirectMemoryAccessStreamReader` class and only implemented the method in the new class to be able to compare them in a single benchmark. I also added a `Reset` method to both so that I could test repeated reads without also measuring the allocations for creating the reader:
```diff
        internal sealed unsafe class DirectMemoryAccessStreamReader : TextReaderWithLength
        {
+           private readonly char* _start;
            private char* _position;
            private readonly char* _end;

            public DirectMemoryAccessStreamReader(char* src, int length)
                : base(length)
            {
                RoslynDebug.Assert(src != null);
                RoslynDebug.Assert(length >= 0);

+                _start = src;
                _position = src;
                _end = _position + length;
            }
+
+           public void Reset()
+           {
+               _position = _start;
+           }
```

</details>

Neme12 commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -468,16 +469,21 @@ public override int Read(char[] buffer, int index, int count)
                     throw new ArgumentNullException(nameof(buffer));
                 }
 
-                if (index < 0 || index >= buffer.Length)
```
Note: I changed this because it was wrong. I should be able to read into a zero-length array when `count` is 0. Or when `index` is equal to `buffer.Length` and `count` is 0.

Also, by .NET convention, if `index + count` is greater than `buffer.Length`, the exception that is thrown shouldn't have `index` or `count` as its parameter. I should be an `ArgumentException`  with a special message without a parameter (because it related to both parameters, or rather all 3 and the relationship between them).

See: https://source.dot.net/#System.Private.CoreLib/StringReader.cs,86

CyrusNajmabadi replied:
thanks for the great explanation :)

Neme12 commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -487,7 +493,36 @@ public override int Read(char[] buffer, int index, int count)
+
+            public override int ReadBlock(Span<char> buffer) =>
+                Read(buffer);
+#endif
```
I have overriden the `ReadBlock` methods as well because the base implementation would use a loop, which isn't needed here. I also provided the span overloads - even though they're not used in `SourceText.From` today, if they were used someday and weren't overriden here, the base implementation in `TextReader` would needlessly allocate an array and forward to the array overloads - it's a fallback back-compat implementation that is really inefficient. In my opinion, the span overloads should always be overriden to avoid that pit of failure, even if we know that the API we're passing the reader into doesn't use them right now.

In general, when implementing something like a `TextReader`, it's always a good idea to implement as many methods as possible and reasonable because we can make it more efficient than the base implementations that have to go through the other ones.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -487,7 +493,36 @@ public override int Read(char[] buffer, int index, int count)
+                return count;
+            }
+
+            public override int ReadBlock(Span<char> buffer) =>
+                Read(buffer);
```
```suggestion
            public override int ReadBlock(Span<char> buffer)
                => Read(buffer);
```

Neme12 replied:
Fixed.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -487,7 +493,36 @@ public override int Read(char[] buffer, int index, int count)
+            public override string ReadToEnd()
+            {
+                if (_position >= _end)
+                    return "";
```
interesting.  is the `> end` case expected?  or can you only have the `==` case?

Neme12 replied:
It shouldn't be needed if everything is correct. I just used to using it likes this defensively. Also, this matches the `count` checks in the other methods where it checks that `count` is greater than zero, as opposed to not equals zero. I can changed this to `==` if you feel that is more appropriate.

CyrusNajmabadi replied:
That would be great.  Can you switch to that, and put a contract call in that _position must be `<=` to _end`?  Thanks!

Neme12 replied:
Fixed.

sharwell commented on src/Dependencies/Collections/Internal/ThrowHelper.cs:
```diff
@@ -279,7 +279,7 @@ private static string GetArgumentName(ExceptionArgument argument)
         }
 
-        private static string GetResourceString(ExceptionResource resource)
+        public static string GetResourceString(this ExceptionResource resource)
```
❗ This is not intended for use outside Microsoft.CodeAnalysis.Collections

Neme12 replied:
@sharwell What should I do instead?

sharwell reviewed (changes requested):
See comment with ❗

sharwell commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -12,6 +12,7 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections.Internal;
```
This API is only intended for use inside Microsoft.CodeAnalysis.Collections

Neme12 commented on src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs:
```diff
@@ -468,16 +468,21 @@ public override int Read(char[] buffer, int index, int count)
 
+                if (index + count > buffer.Length)
+                {
+                    throw new ArgumentException();
```
Is this OK then since this isn't a public API?

Neme12 commented:
@sharwell Could you please help me understand what the failures might be? It says there were test failures, for example here: https://dev.azure.com/dnceng/public/_build/results?buildId=1774387&view=logs&jobId=b62c9f97-affa-59fc-36bf-3106c2ea085c&j=b62c9f97-affa-59fc-36bf-3106c2ea085c&t=feacc3f4-7491-5282-b8ee-3a66ef7022bb

I don't get any test results in there, but there's a link to a failure log:
![image](https://user-images.githubusercontent.com/11444821/168704592-ad86e0ea-5c9a-4ef8-8155-e98da38f872d.png)

When I open that however (https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-roslyn-refs-pull-61351-merge-c8f5ae474c46418ca2/Microsoft.CodeAnalysis.Workspaces.UnitTests.dll.3.Attempt.3/1/console.4f70f924.log?helixlogtype=result), there are no test failures there either. The only thing I can find about test results in that file is this:
![image](https://user-images.githubusercontent.com/11444821/168704742-445873ba-0d88-41d1-af2e-453c92ec4d33.png)

It says 0 failed, 49 passed. That number seems really low though. Can I found the actual failures somewhere? The test I added pass locally for me.

Neme12 commented:
I'm wondering if this could be related to the fact that I added `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>` to the test project, or that I added a test that is only compiled on .NET Core.

Neme12 commented:
I also says "Failed to upload test results".
![image](https://user-images.githubusercontent.com/11444821/168705693-3a9a2cc7-975f-414c-90f6-69eebaf3027e.png)

Neme12 commented:
@sharwell I added a workaround the CI issue (the integration test failures are unrelated and can be seen in other PRs as well). Is this good to merge now?

Neme12 commented:
@sharwell The CI issues are fixed. Could you please make another quick look at some of the test changes I made since?

sharwell commented on src/Workspaces/CoreTest/WorkspaceServiceTests/DirectMemoryAccessStreamReaderTests.cs:
```diff
@@ -0,0 +1,160 @@
+        }
+
+        [Theory]
+        [InlineData(5, 0, 5, "bcdef", 5, 'g')]
+        [InlineData(5, 0, 2, "bc\0\0\0", 2, 'd')]
+        [InlineData(10, 1, 6, "\0bcdefg\0\0\0", 6, 'h')]
+        [InlineData(10, 1, 7, "\0bcdefgh\0\0", 7, -1)]
+        [InlineData(10, 1, 9, "\0bcdefgh\0\0", 7, -1)]
+        [InlineData(10, 3, 7, "\0\0\0bcdefgh", 7, -1)]
+        [InlineData(10, 3, 0, "\0\0\0\0\0\0\0\0\0\0", 0, 'b')]
+        [InlineData(10, 10, 0, "\0\0\0\0\0\0\0\0\0\0", 0, 'b')]
+        [InlineData(0, 0, 0, "", 0, 'b')]
+        public void ReadToArray(int bufferLength, int index, int count, string expected, int expectedResult, int expectedPeek)
```
💡 In a situation like this, I would typically make two changes:

1. I would remove all of the "expected" parameters. This causes the test to be parameterized only on the inputs, and if the outputs ever change the test signature itself doesn't change and therefore the CI history preserves the long-term test statistics.
2. Instead of hard-coding the expected outputs, I would create a reference stream instance alongside the one we are testing (e.g. `MemoryStream`). Then instead of verifying specific values, I would assert that both streams produce the same result from the same API calls.

Neme12 replied:
@sharwell  Ah, I see what you mean. Comparing the behavior to a built-in `TextReader` like `StringReader` or `StreamReader`? The expected outputs shouldn't ever change and I think it's useful to see the expected outputs in the test code like this so that we can easily see what sorts of interesting cases are tested, including their outputs, and whether there's an interesting case that we're missing.

CyrusNajmabadi replied:
I think the point here is that the outputs aren't relevant (and on their own don't tell us if this is correct or not).  What we want to show is that this type behaves identically to any other stream-reader type in terms of semantics.  so as long as they all agree on the behavior, then this is a suitable type to use.

Neme12 replied:
@CyrusNajmabadi Why are they not relevant? I think the outputs are really useful so that you can see 1) whether the behavior is correct 2) what sorts of interesting case we're testing, including the various outputs we're testing. If it was only compared to a built-in `TextReader`, you wouldn't be able to see what sorts of different kinds of outputs we're testing.

Neme12 replied:
Also, the inputs are chosen partially based on the different outputs we want to get and test. If you don't see the output, you can't compare the inputs to the outputs easily and see why some of these particular inputs are chosen.

Neme12 replied:
Would it be OK if I added an additional test that compares to an existing reader without getting rid of the tests that use expected outputs?

CyrusNajmabadi replied:
> @CyrusNajmabadi Why are they not relevant? I think the outputs are really useful so that you can see 1) whether the behavior is correct

Because i woudl have to figure out what correctness means.  Which means i could make a mistake there.  I trust the battle-tested shipping impls of this APIs more than me.  So i'd rather compare the results to *them* rather to me :)

Neme12 commented on src/Workspaces/CoreTest/WorkspaceServiceTests/DirectMemoryAccessStreamReaderTests.cs:
```diff
@@ -0,0 +1,160 @@
+            var handle = GCHandle.Alloc(text.ToCharArray(), GCHandleType.Pinned);
+            reader = new DirectMemoryAccessStreamReader((char*)handle.AddrOfPinnedObject(), text.Length);
+
+#pragma warning disable RS0042 // Do not copy value
```
Fixed in c7b7e5d0951e1a671dbe28c2d61148f7cd7b410d

Neme12 commented on src/Workspaces/CoreTestUtilities/TextReaderTestBase.cs:
```diff
@@ -0,0 +1,167 @@
+            AssertAllEqual('e', referenceReader.Read(), reader.Read());
+            AssertAllEqual('x', referenceReader.Read(), reader.Read());
+            AssertAllEqual('t', referenceReader.Read(), reader.Read());
+            AssertAllEqual(-1, referenceReader.Peek(), reader.Peek());
```
I get that testing against a reference `TextReader` implementation is useful, but basically I don't like the idea of removing the expected values, because then you can't see if we're testing reading at the end for example to get -1. If these lines only contained checking against the reference implementation, you wouldn't see whether we're testing interesting cases like that.

Neme12 commented:
@sharwell Is this OK now?
